File: sherry/week-3/464. Sort Integers II.java
// 464. Sort Integers II
//
// Given an integer array, sort it in ascending order. Use quick sort, merge sort, heap sort or any O(nlogn) algorithm.
//
// Example
//
// Given [3, 2, 1, 4, 5], return [1, 2, 3, 4, 5].
// TODO; using merge sort
public class Solution {
    /**
     * @param A: an integer array
     * @return: nothing
     */
    public void sortIntegers2(int[] A) {
        // write your code here
        // edge cases
        if(A == null || A.length == 0) {
            return;
        }
        quickSort(A, 0, A.length-1);

    }
    private void quickSort(int[] nums, int start, int end) {
        if (start >= end) {
            return;
        }
        int left = start;
        int right = end;
        int pivot = nums[left + (right - left) / 2];
        while (left <= right) {
            while (left <= right && nums[left] < pivot) {
                left++;
            }
            while (left <= right && nums[right] > pivot) {
                right--;
            }
            //swap
            if (left <= right) {
                int temp = nums[left];
                nums[left] = nums[right];
                nums[right] = temp;
                left++;
                right--;
            }
        }
        quickSort(nums, start, right);
        quickSort(nums, left, end);
    }
}

File: Joe Zheng/week 3/464. Sort Integers II.py
class Solution:
    """
    @param A: an integer array
    @return: nothing
    """

    def sortIntegers2(self, A):
        # write your code here
        # quickSort

        self.quickSort(0, len(A) - 1, A)

    def quickSort(self, start, end, A):
        if start >= end:
            return

        left, right = start, end
        pivot = A[(start + end) // 2]

        while left <= right:
            if A[left] < pivot:
                left += 1
                continue
            if A[right] > pivot:
                right -= 1
                continue

            # if left <= right:
            A[left], A[right] = A[right], A[left]
            left += 1
            right -= 1

        # sort left then right
        # the only lapping part does not need to sort
        self.quickSort(start, right, A)
        self.quickSort(left, end, A)
